이모지를 채널에 날릴 수 있다!🤭

happydhKim opened the pull request:
# [Epic#38] 이모지를 채널에 날릴 수 있다!🤭


## 해당 이슈 📎

- [Epic#38] 이모지를 채널에 날릴 수 있다!🤭(#87)


## 변경 사항 🛠

이모지를 채널에 날리며, 모두 이모지를 확인할 수 있습니다.

- 전체화면에서는 이모지의 크기가 커집니다.
- 전체화면에서 이모지는 우측하단에서 나옵니다.

## 테스트 ✨

> 없음

## 리뷰어 참고 사항 🙋‍♀️

- 현재 자원을 많이 사용하는 방식으로 되어있습니다.
- 개선사항이나 코드를 더 깔끔하게 할 수 있는 부분을 중점적으로 봐주시면 감사하겠습니다.😀

happydhKim committed:
[ FEATURE ] 날아가는 이모지의 출발점인 버튼을 구현 (#87)

Why
- 날아가는 이모지를 선택할 수 있는 버튼이 필요하다.

What
- popover를 나타낼 수 있는 버튼 구현.
- popover에 총 5가지의 이모지를 임시로 구현.

happydhKim committed:
Merge branch 'master' of https://github.com/connect-foundation/2019-02 into feature/issue87

happydhKim committed:
[ FEATURE ] 날아가는 이모지 컴포넌트 생성 (#87)

Why
- 버튼이 눌리면 날아가는 이모지를 담는 공간이 필요하다.

What
- 날아가는 이모지 컴포넌트 생성.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 초기 구현 완료 (#87)

Why
- 이모지 버튼을 누르면 이모지가 날아가야 한다.

What
- 이모지 초기 날아가는 애니메이션 구현 완료.
- 초기에 버튼을 누르지 않아도 렌더링이 됨에 따라 자동 날라가는 현상 있음.
- popover가 나오면 바로 나오는 현상 발생.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 구현 (#87)

Why
- 스피커와 리스너 간 소통을 도와주기 위해 날아가는 이모지가 필요하다.

What
- 감정표현 버튼 을 누른 후에 이모지가 팝오버로 나오게 구현.
- 팝오버에 있는 이모지를 클릭 시에 이모지가 날아가도록 구현.
- 이모지가 상단으로 위치 할 시, 사라지도록 구현.
- 이모지가 모두 날아 간다면, cancelAnimationFrame을 통해 action을 중단하도록 구현.

happydhKim committed:
[ REFACTOR ] 불필요한 코드 제거 (#87)

Why
- 불필요한 코드가 있어 수정이 필요하다.

What
- 불필요한 코드 제거.

happydhKim committed:
[ MERGE ] 최신버전으로 업데이트 (#87)

Why
- 최신버전의 코드로 업데이트 해야 한다.

What
- 최신버전의 devleop 브랜치와 병합.

happydhKim committed:
[ REFACTOR ] 기존 popver방식을 일반 버튼 나열로 변경 (#87)

Why
- 기존의 popover방식은 번거로운 방식이라 변경해야 한다.

What
- 바로 이모지를 날릴 수 있도록 버튼에 나열하는 방식으로 변경.
- 기존의 popover 삭제.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 출발 위치를 버튼으로 변경 (#87)

Why
- 날아가는 이모지의 출발점이 중앙 맨 하단이라 어색함이 있다.
- 스크린을 가리는 현상이 발생한다.
- 이모지의 크기가 너무 작다.

What
- 날아가는 이모지의 출발점을 버튼 위치로 변경했다.
- 이모지의 크기를 약간 키웠다.
- 이모지가 시간이 지날 수록 투명화 되며 없어진다.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 동기화 작업 : 스키마 작성 (#87)

Why
- 날아가는 이모지 동기화를 위해 스키마를 작성해야 한다.

What
- 데이터베이스를 이용하지 않고, 이벤트 기반으로 작업을 진행 하도록 설계하였다.
- typedefs, resolvers, hook을 각각 설계.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 동기화 구현 (#87)

Why
- 날아가는 이모지를 동기화 시켜야 한다.
- 동기화를 시키기 위해 구독 및 스키마를 설정해야 한다.
- 이모지를 날리기 위해 출발 좌표와 채널 그리고 어떤 이모지인지 확인할 수 있어야 한다.

What
- 구독 스키마를 설계했다.
- 구독 스키마를 사용하기 위해 useCreateEmoji, useAddEmoji 훅을 코딩했다.
- 이모지를 날리기 위해 출발 좌표 그리고 type으로 어떤 이모지인지 확인할 수 있게 만들었다.
- 이모지를 날리기 위해 useCreateEmoji를 확인해 출발시킨다.

happydhKim committed:
[ MERGE ] 최신 버전으로 업데이트 (#87)

Why
- 코드를 최신 버전으로 동기화를 시켜야 한다.

What
- 최신버전의 develop 코드를 가지고 온다.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 전체화면 여부 상태값 전달 (#87)

Why
- 일반화면과 전체화면에서 날아가는 이모지의 크기 및 출발 위치가 틀려야 한다.

What
- 일반화면과 전체화면을 구분 지을 수 있는 isFullScreen 상태 값을 전달.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 전체화면 구현 (#87)

Why
- 날아가는 이모지가 전체화면에서도 동작해야 한다.

What
- 날아가는 이모지 전체화면에서도 동작하도록 구현.
- isFullScreen 변수를 Factory에서 받아 구별할 수 있도록 한다.

happydhKim committed:
[ FEATURE ] 날아가는 이모지 전체 화면일때 크기 조절 (#87)

Why
- 전체화면에서는 이모지가 약간 더 커야 한다.

What
- 일반화면과 전체화면에서의 이모지의 크기를 구별 할 수 있도록 구현.

happydhKim committed:
[ REFACTOR ] 날아가는 이모지 버튼 수정 (#87)

Why
- 날아가는 이모지를 간소화 하고 아바타 형식으로 바꾼다.
- 이모지가 너무 많으면 집중을 저하시킬 수 있다.
- 깔끔하게 하기 위해 아바타 형식으로 바꾼다.

What
- 이모지를 가장 필요한 조건으로 변경.
- 우울한 이모지는 발표자의 기분을 상하게 할 수 있기 때문에 제외시킨다.
- 이모지를 아바타 형식으로 감싸줬다.

happydhKim committed:
[ REFACTOR ] 코드 개선 (#87)

Why
- 린트에 걸린 코드를 수정한다.

What
- 코드 수정

aereeeee commented on web/src/components/channel/Slide/Slide.jsx:
```diff
@@ -19,12 +19,16 @@ const Slide = (props) => {
 };
 
 Slide.propTypes = {
+  channelId: PropTypes.string.isRequired,
```
develop브랜치 머지하고 다시 PR 날려주세요!

happydhKim replied:
지금 받았어여 😃

happydhKim committed:
[ MERGE ] 최신버전으로 업데이트 (#87)

Why
- PR을 날리기 전 최신버전으로 업데이트를 해야 한다.

What
- 최신버전으로 업데이트

aereeeee commented on web/src/components/channel/Slide/SlideInfo/FlyingEmojiButton/FlyingEmojiButton.jsx:
```diff
@@ -0,0 +1,83 @@
+    if (isReadyBroadcastData()) return;
+    const { type, positionX, positionY } = broadcastEmoji;
+    const fullScreenPositionX = body.offsetWidth - 150;
+    const getFlyingEmojiSpeed = 3 + Math.random() * 8;
```
매직넘버가 많은것 같아요 왜 저런 숫자가 들어갔는지 코드로 이해하기 힘들어서 변수로 할당하는게 더 좋을것 같습니다!

happydhKim replied:
넵 constants로 뺐습니다!! 🐶

aereeeee commented on web/src/components/channel/Slide/SlideInfo/FlyingEmojiButton/FlyingEmojiButton.jsx:
```diff
@@ -0,0 +1,83 @@
+    const getFlyingEmojiSpeed = 3 + Math.random() * 8;
+    const samePosition = isFullScreen
+      ? { x: fullScreenPositionX, y: 0 }
+      : { x: positionX, y: positionY };
```
컨벤션에 맞게 선언문 다음에 개행해주세요~!

happydhKim replied:
넹😥

aereeeee commented on web/src/components/channel/Slide/SlideInfo/FlyingEmojiButton/FlyingEmojiButton.jsx:
```diff
@@ -0,0 +1,83 @@
+
+  return (
+    <S.EmojiSmallButton>
+      <S.EmojiButton onClick={(e) => emojiMaker(e, '❤️')}>
```
e말고 event로 쓰기로 하였으니 수정해주세요~!

happydhKim replied:
넵 수정했어요🐤

aereeeee commented on web/src/components/channel/Slide/SlideInfo/FlyingEmojiFactory/FlyingEmojiFactory.jsx:
```diff
@@ -0,0 +1,81 @@
+    this.coordinates = coordinates;
+    this.position = -(this.coordinates.y);
+    this.isFullScreen = isFullScreen;
+    this.scale = isFullScreen ? 4.5 + Math.random() : 3 + Math.random();
```
모듈에도 매직넘버가 많은것 같네요! 얘를 들어서 scaleOffest같이 인자로 offset값을 받아서 상대값으로 4.5와 3을 대체하면 어떨까요? 좀더 유틸로서 역할이 확실해 질것 같아요

happydhKim replied:
넵 복잡한 연산이 있는 녀석들을 constants로 뺐습니당!

aereeeee commented on web/src/components/channel/Slide/SlideInfo/FlyingEmojiFactory/FlyingEmojiFactory.jsx:
```diff
@@ -0,0 +1,81 @@
+  flying() {
+    this.getPosition();
+    this.setOpacity();
+    const height = this.position;
```
선언문과 실행문 사이 개행해주세요~

happydhKim replied:
넹😥

aereeeee commented on web/src/components/channel/Slide/SlideInfo/FlyingEmojiFactory/FlyingEmojiFactory.jsx:
```diff
@@ -0,0 +1,81 @@
+  render() {
+    const background = document.querySelector('body');
+    const fullScreen = document.querySelector('.fullscreen');
+    this.item = this.emoji;
```
여기도 개행이요~

happydhKim replied:
넹😥

aereeeee commented on web/src/components/channel/Slide/SlideInfo/SlideInfo.jsx:
```diff
@@ -1,10 +1,13 @@
 
-const SlideInfo = () => {
+const SlideInfo = (props) => {
+  const { channelId, isFullScreen } = props;
```
channelId 상태 context에 등록해 놓아서 useChannelSelector로 사용할 수 있어요.
isfullScreen도 채널 리듀서에 등록하여 전역으로 사용하면 슬라이드 컴포넌트에 props를 없앨수있을 것 같아요.

aereeeee reviewed (approved):
드디어 플라잉 이모지 이펙트를 완성했네요 정말 멋집니다 👍 👍

happydhKim committed:
[ REFACTOR ] 날아가는 이모지 PR 리뷰 받은 부분 수정 (#87)

Why
- 리뷰 받은 부분을 수정해야 한다.
- 큰 이슈로 매직넘버를 관리할 수 있는 부분이 필요하다.

What
- constants를 생성하여, 매직넘버를 관리한다.
- 부분 지적받은 부분을 수정.
